### DISCUSSION TITLE
Contribution

### DISCUSSION THREAD

**Discussion by ArishSultan** [isar/isar#17] [Help]
Hi, I hope you are doing fine. I just saw this project and I liked the idea, I want to work on it just let me know if there are some tasks or if you have a specific contribution guideline, etc.

**Comment by simc** [comment]
Hey, I don't have a contribution guide yet but I appreciate any help 👍 
Would you like to contribute to isar (Dart) or isar-core (Rust)?

**Reply by ArishSultan** [reply]
I would like to contribute to isar (Dart) for now.

since you said that isar-core is stable enough so, i would like to work on future plans.

**Reply by simc** [reply]
There is a lot to do on the Dart side:
- Unit tests
- An extensive demo app
- Examples
- API docs
- later: a tool to remotely inspect the database

Give me one or two days, I'm working on major changes (already implemented in core). Then you can decide what you want to work on ;)

**Reply by simc** [reply]
The new release is out :)

**Comment by AKushWarrior** [comment]
I'd also be interested. I can contribute to either the Dart interface or the Rust base.

**Reply by simc** [reply]
Core is already quite stable and should not change too much. Right now the most work is writing good unit tests in Rust. All files in Core need unit tests. The difficulty is that the Core API is quite verbose (by design) which makes it a little harder to write small and easy to read unit tests. I created [some macros](https://github.com/isar/isar-core/blob/main/src/utils/debug.rs) to make opening Isar instances easier.

**Reply by simc** [reply]
@AKushWarrior It would also be amazing if you could maybe start with a web Implementation 😉

**Reply by AKushWarrior** [reply]
I can give it a go. Should I model the API off the Dart interface? Also, do you prefer ES6 or TypeScript?

**Reply by simc** [reply]
I think TypeScript would be best. Yes, the Dart API (and also Core) are designed to work with IndexedDB. Synchronous operations cannot be ported to web.
I would start with CRUD operations `get()`, `getAll()`, `put()` etc.

We should store objects as JavaScript objects (Maps in Dart). We should use the isar_generator to generate `toIsar()` and `fromIsar()` methods to convert Dart objects to / from maps.

This is a huge task. You can also choose something else if you like.

**Reply by AKushWarrior** [reply]
Hey, I'm starting to work on this. Would you mind initializing an empty git repo in https://github.com/isar/isar-web so I can file PRs?

**Comment by simc** [comment]
Here are the next tasks: https://github.com/isar/isar/blob/main/TODO.md

**Reply by DevNico** [reply]
I'd like to contribute by creating example apps for isar. Is the API stable as of right now e.g. does it even make sense to write example apps already?

I was thinking about a todo app as the minimal example and then to make it more advanced implement full text seach as well as tagging and multiple todo lists.

I could also port all the examples from hive to isar if that's something you want.

**Reply by simc** [reply]
@DevNico That would be great. The latest version on pub.dev has an outdated API but I will publish a new version with updated documentation very shortly. I'll ping you here when I'm finished.

>  was thinking about a todo app as the minimal example and then to make it more advanced implement full text seach as well as tagging and multiple todo lists.

That sounds definitely cool. Maybe you can somehow use the new `IsarLink` and `IsarLinks` to demonstrate relationships.

It would also be great to have a second app some time in the future that shows what isar is capable of in terms of performance (a few 100k entries should be no problem at all)

And if you like, you can document your app or add some guide and we can publish it as part of the docs (like [hive tutorials](https://docs.hivedb.dev/#/tutorials/favorite_books))

**Reply by DevNico** [reply]
Sounds good, I'll definitely write a guide. I was thinking about publishing it to medium and similar platforms as well then if that's alright with you.

I'll think about how to best show of IsarLink in an example that makes sense.

**Reply by leonfuss** [reply]
If you want to stick to a ToDo app, you could just allow creating subtasks that are linked to the parent task

**Reply by simc** [reply]
@DevNico New version is published, updated docs will follow soon 🙌

**Reply by simc** [reply]
Docs are also updated
